util/slack_wrapper.py

```diff
@@ -1,202 +1,223 @@
+
+    @property
+    def connected(self):
+        return self.socket.is_connected()
+
+    def handle(self, client: SocketModeClient, message: SocketModeRequest):
+        # send the response immediately so we don't get retransmissions
+        client.send_socket_mode_response(SocketModeResponse(message.envelope_id))
 
-        if self.connected:
-            self.server = self.client.server
-            self.username = self.server.username
-            self.user_id = self.server.login_data.get("self").get("id")
+        # slack is super eager to retransmit so here's a free memory leak
+        id = message.payload['event']['client_msg_id']
+        if id in self.seen_messages:
+            return
 
-    def read(self):
-        """Read from the real-time messaging API."""
-        return self.client.rtm_read()
+        self.seen_messages.append(id)
 
-    def invite_user(self, users, channel, is_private=False):
+        for handler in self.on_message_handlers:
+            try:
+                handler(message)
+            except Exception as e:
+                import traceback
+                traceback.print_exc()
+                print("error handling message", e)
+
+    def on_message(self, handler):
+        self.on_message_handlers.append(handler)
+
+    def invite_user(self, user: Union[str, Sequence[str]], channel: str) -> Dict:
         """
         Invite the given user(s) to the given channel.
         """
 
-        users = [users] if not type(users) == list else users
-        api_call = "conversations.invite"
-        return self.client.api_call(api_call, channel=channel, users=users)
+        return self.client.conversations_invite(channel=channel, users=user).data
 
-    def set_purpose(self, channel, purpose, is_private=False):
+    def set_purpose(self, channel: str, purpose: str) -> Dict:
         """
         Set the purpose of a given channel.
         """
 
-        api_call = "conversations.setPurpose"
-        return self.client.api_call(api_call, purpose=purpose, channel=channel)
+        return self.client.conversations_setPurpose(channel=channel, purpose=purpose).data
 
-    def set_topic(self, channel, topic, is_private=False):
+    def set_topic(self, channel: str, topic: str) -> Dict:
         """Set the topic of a given channel."""
 
-        api_call = "groups.setTopic" if is_private else "channels.setTopic"
-        return self.client.api_call(api_call, topic=topic, channel=channel)
+        return self.client.conversations_setTopic(channel=channel, topic=topic).data
 
-    def get_members(self):
+    def get_members(self) -> List[Dict]:
         """
-        Return a list of all members.
+        Return a list of all members. Will raise exception if list fails
         """
-        return self.client.api_call("users.list", presence=True)
 
-    def get_member(self, user_id):
+        members = []
+        for page in self.client.users_list():
+            members.extend(page['members'])
```

Comment:
`return [page["members"] for page in self.client.users_list()]`

Reply:
Wouldn't this just return a list of lists? I suppose we could `return [member for member in page["members"] for page in self.client.users_list()]` but I'm not sure how readable that is

Reply:
you're right, my bad :)